### ISSUE TITLE
grpc-web does not work with akka-gRPC 2.1.0 / akka-http 10.2.6

### ISSUE THREAD

**Issue by kallependers** [akka/akka-grpc#1459]
After updating akka-http from 10.2.4 to 10.2.6 and akka-grpc from 2.0.0 to 2.1.0 gRPC web requests (accept: application/grpc-web-text) are failing.

- no errors in the logs
- the gRPC request succeeds
- the response fails

So I guess it goes somewhere wrong when generating the proper response for the accept.

**Comment by jrudolph** [comment]
Moved to akka-grpc.

Thanks for the report, @kallependers. Interesting that this wouldn't be tested.

**Comment by jrudolph** [comment]
Could you post the full output of using `curl -v` to test this issue?

**Comment by kallependers** [comment]
```
$ curl -v 'https://example.com/grpc/Endpoint' \
-H 'authority: example.com' \
-H 'sec-ch-ua: "Google Chrome";v="93", " Not;A Brand";v="99", "Chromium";v="93"' \
-H 'x-user-agent: grpc-web-javascript/0.1' \
-H 'sec-ch-ua-mobile: ?0' \
-H 'authorization: Bearer JWT' \
-H 'content-type: application/grpc-web-text' \
-H 'accept: application/grpc-web-text' \
-H 'x-grpc-web: 1' \
-H 'user-agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/93.0.4577.82 Safari/537.36' \
-H 'sec-ch-ua-platform: "Linux"' \
-H 'origin: https://example.com' \
-H 'sec-fetch-site: same-site' \
-H 'sec-fetch-mode: cors' \
-H 'sec-fetch-dest: empty' \
-H 'accept-language: en-US,en;q=0.9,nl;q=0.8' \
--data-raw 'AAAAAAA=' \
--compressed
*   Trying IP:443...
* Connected to example.com (IP) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*  CAfile: /etc/ssl/certs/ca-certificates.crt
*  CApath: /etc/ssl/certs
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
* TLSv1.3 (IN), TLS handshake, Finished (20):
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.3 (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384
* ALPN, server accepted to use h2
* Server certificate:
*  subject: CN=example.com
*  start date: Aug  2 09:34:05 2021 GMT
*  expire date: Oct 31 09:34:03 2021 GMT
*  subjectAltName: host "example.com" matched cert's "example.com"
*  issuer: C=US; O=Let's Encrypt; CN=R3
*  SSL certificate verify ok.
* Using HTTP2, server supports multi-use
* Connection state changed (HTTP/2 confirmed)
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* Using Stream ID: 1 (easy handle 0x55abac9b5580)
> POST /grpc/Endpoint HTTP/2
> Host: example.com
> accept-encoding: deflate, gzip, br
> authority: example.com
> sec-ch-ua: "Google Chrome";v="93", " Not;A Brand";v="99", "Chromium";v="93"
> x-user-agent: grpc-web-javascript/0.1
> sec-ch-ua-mobile: ?0
> authorization: Bearer JWT
> content-type: application/grpc-web-text
> accept: application/grpc-web-text
> x-grpc-web: 1
> user-agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/93.0.4577.82 Safari/537.36
> sec-ch-ua-platform: "Linux"
> origin: https://example.com
> sec-fetch-site: same-site
> sec-fetch-mode: cors
> sec-fetch-dest: empty
> referer: https://example.com/
> accept-language: en-US,en;q=0.9,nl;q=0.8
> content-length: 8
>
* We are completely uploaded and fine
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* old SSL session ID is stale, removing
* Connection state changed (MAX_CONCURRENT_STREAMS == 2147483647)!
  < HTTP/2 200
  < content-type: application/grpc-web-text+proto
  < content-length: 128
  < trace-id: TRACE_ID
  < grpc-encoding: identity
  < x-request-id: REQUEST_ID
  < date: Thu, 23 Sep 2021 12:03:39 GMT
  < server: istio-envoy
  < x-envoy-upstream-service-time: 15
  < strict-transport-security: max-age=31536000; includeSubDomains
  < access-control-allow-origin: https://example.com
  < access-control-allow-credentials: true
  < access-control-expose-headers: custom-header-1,grpc-message,grpc-status,x-request-id
  <
* transfer closed with 128 bytes remaining to read
* stopped the pause stream!
* Connection #0 to host example.com left intact
  curl: (18) transfer closed with 128 bytes remaining to read
```
This call is made on a staging environment.

**Comment by jrudolph** [comment]
So, you get a successful `200` response, but the response is never delivered?

**Comment by kallependers** [comment]
Correct

**Comment by drmontgomery** [comment]
My team is also experiencing issues with grpc-web after upgrading to akka-gRPC 2.1.0. We're using grpc-web-text to make requests from a browser-based single page app directly to an Akka HttpServer. We've identified two specific problems.

First, there appears to be a bug in `AbstractGrpcProtocol.reader` that causes valid requests to generate errors if they take the strict path in `GrpcMarshalling.unmarshal`. Looking at it in the debugger, the `decoder` function defined in `AbstractGrpcProtocol.reader` reads invalid data for `frameType` and `length`, which causes `reader.take(length)` to fail with an exception.

This bug can be fixed by applying `strictAdapter` to the entire frame (i.e. prior to creating the ByteReader), rather than just the contents of data frames. This aligns with the order in `GrpcProtocolWeb.encodeFrame` (compress, add framing data, and then postEncode with base64.) I've tested this change locally, and it solves the problem.

After fixing the unmarshalling problem, however, we have encountered a separate issue in the response rendering. We're getting the following exception:
```
java.lang.IllegalStateException: Unexpected protocol 'HttpProtocol(HTTP/2.0)'
	at akka.http.impl.engine.rendering.HttpResponseRendererFactory$HttpResponseRenderer$$anon$1.renderStatusLine$1(HttpResponseRendererFactory.scala:142)
	at akka.http.impl.engine.rendering.HttpResponseRendererFactory$HttpResponseRenderer$$anon$1.akka$http$impl$engine$rendering$HttpResponseRendererFactory$HttpResponseRenderer$$anon$$render(HttpResponseRendererFactory.scala:293)
	at akka.http.impl.engine.rendering.HttpResponseRendererFactory$HttpResponseRenderer$$anon$1$$anon$2.onPush(HttpResponseRendererFactory.scala:63)
```

This appears to be happening because `GrpcResponseHelpers.responseWithTrailers` creates an HttpResponse with an explicit HTTP/2.0 protocol, which doesn't play well with the HTTP/1.1 request/response flow for grpc-web.

I'm not sure the best way to fix this. AbstractGrpcProtocol/GrpcProtocolWeb are both agnostic with respect to HTTP protocol. It seems like this could potentially be handled by the generated handler code for each service, either by passing the protocol from the request into GrpcMarshalling.marshal(), or calling withProtocol on the generated HttpResponse. Both of those options feel a little ugly, though, so I'm hoping that there is a better solution.

Please let me know if you'd like me to submit a pull request for the decoder fix.

**Comment by drmontgomery** [comment]
I suppose another option is to modify GrpcResponseHelper.responseWithTrailers to use HttpProtocol.Http/1.1, like it did in akka-gRPC v2.0.0. I've tested this locally, and it solves the problem for our use case and doesn't appear to break native gRPC.

**Comment by raboof** [comment]
Thanks for digging in! These two improvements *sound* OK to me, so if you can PR and CI is agrees I'd be happy to accept them!

**Comment by jrudolph** [comment]
Should be fixed by #1462.
